File: src/main/java/frc/robot/subsystems/Powertrain.java
package frc.robot.subsystems;

import edu.wpi.first.wpilibj2.command.SubsystemBase;
import edu.wpi.first.wpilibj.drive.DifferentialDrive;
import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;

import com.revrobotics.CANSparkMax;
import com.revrobotics.CANSparkMaxLowLevel.MotorType;

import frc.robot.Constants.drivetrainConstants;

import static frc.robot.subsystems.Powertrain.Mode.*;

public class Powertrain extends SubsystemBase {
    public final CANSparkMax left1 = new CANSparkMax(drivetrainConstants.motors_left[0], MotorType.kBrushless);
    public final CANSparkMax left2 = new CANSparkMax(drivetrainConstants.motors_left[1], MotorType.kBrushless);
    public final CANSparkMax right1 = new CANSparkMax(drivetrainConstants.motors_right[0], MotorType.kBrushless);
    public final CANSparkMax right2 = new CANSparkMax(drivetrainConstants.motors_right[1], MotorType.kBrushless);

    public final CANSparkMax[] motors = {left1, left2, right1, right2};

    public final MotorControllerGroup left_motors = new MotorControllerGroup(left1, left2);
    public final MotorControllerGroup right_motors = new MotorControllerGroup(right1, right2);
    
    public final DifferentialDrive differentialDrive = new DifferentialDrive(left_motors, right_motors);

    private double power;
    private double rotation;
    private double leftPower;
    private double rightPower;

    public Mode mode;

    public Powertrain() {
        for (CANSparkMax motor : motors) {
            ConfigureSpark(motor);
        }
        reset();
    }

    private void ConfigureSpark(CANSparkMax motor) {
        motor.enableVoltageCompensation(drivetrainConstants.voltage_compensation);
        motor.setSmartCurrentLimit(drivetrainConstants.stall_current_limit);
        motor.setIdleMode(drivetrainConstants.default_mode);
        motor.setClosedLoopRampRate(drivetrainConstants.ramp_rate);
        motor.setInverted(drivetrainConstants.reverse_motor);
        motor.burnFlash();
    }

    public void reset() {
        mode = CURVATURE_DRIVE;

        power = 0;
        rotation = 0;
        leftPower = 0;
        rightPower = 0;
    }

    private void setTankPowers(double left_power, double right_power) {
        mode = TANK_DRIVE;
        power = rotation = 0;

        this.leftPower = left_power;
        this.rightPower = right_power;
    }
    
    private void setArcadePowers(double power, double rotation) {
        mode = ARCADE_DRIVE;
        leftPower = rightPower = 0;

        this.power = power;
        this.rotation = rotation;
    }

    public void tankDrive(double left_power, double right_power) {
        mode = TANK_DRIVE;
        setTankPowers(left_power, right_power);
    }
    public void arcadeDrive(double power, double rotation) {
        mode = ARCADE_DRIVE;
        setArcadePowers(power, rotation);
    }
    public void curvatureDrive(double power, double rotation) {
        mode = CURVATURE_DRIVE;
        setArcadePowers(power, rotation);
    }

    @Override
    public void periodic() {
        if (mode == TANK_DRIVE){
            differentialDrive.tankDrive(leftPower, rightPower, false);
        }
        else if (mode == ARCADE_DRIVE){
            differentialDrive.arcadeDrive(power, rotation, false);
        }
        else if (mode == CURVATURE_DRIVE){
            differentialDrive.curvatureDrive(power, rotation, true);
        }
    }

    enum Mode {
        TANK_DRIVE(0),
        CURVATURE_DRIVE(1),
        ARCADE_DRIVE(2);

        final int value;
        Mode(int value) {
            this.value = value;
        }
    }
}

File: src/main/java/frc/robot/subsystems/Powertrain.kt
package frc.robot.subsystems

import com.revrobotics.CANSparkMax
import com.revrobotics.CANSparkMax.IdleMode
import com.revrobotics.CANSparkMaxLowLevel
import edu.wpi.first.wpilibj.drive.DifferentialDrive
import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup
import edu.wpi.first.wpilibj2.command.SubsystemBase

class Powertrain : SubsystemBase() {
    val left1 = CANSparkMax(1, CANSparkMaxLowLevel.MotorType.kBrushless)
    val left2 = CANSparkMax(2, CANSparkMaxLowLevel.MotorType.kBrushless)
    val right1 = CANSparkMax(3, CANSparkMaxLowLevel.MotorType.kBrushless)
    val right2 = CANSparkMax(4, CANSparkMaxLowLevel.MotorType.kBrushless)
    val motors = listOf(left1, left2, right1, right2)
        .onEach(::configureSpark)
    val differentialDrive = DifferentialDrive(
        MotorControllerGroup(left1, left2),
        MotorControllerGroup(right1, right2)
    )

    private var power = 0.0
    private var rotation = 0.0
    private var leftPower = 0.0
    private var rightPower = 0.0

    var mode: Mode? = null

    init {
        reset()
    }

    private fun configureSpark(motor: CANSparkMax) {
        motor.enableVoltageCompensation(11.0)
        motor.setSmartCurrentLimit(39)
        motor.idleMode = IdleMode.kBrake
        motor.closedLoopRampRate = 1.0
        motor.inverted = true
        motor.burnFlash()
    }

    fun reset() {
        mode = Mode.CURVATURE_DRIVE
        power = 0.0
        rotation = 0.0
        leftPower = 0.0
        rightPower = 0.0
    }

    fun tankDrive(leftPower: Double, rightPower: Double) {
        mode = Mode.TANK_DRIVE
        setTankPowers(leftPower, rightPower)
    }

    private fun setTankPowers(leftPower: Double, rightPower: Double) {
        rotation = 0.0
        power = rotation
        this.leftPower = leftPower
        this.rightPower = rightPower
    }

    fun arcadeDrive(power: Double, rotation: Double) {
        mode = Mode.ARCADE_DRIVE
        setArcadePowers(power, rotation)
    }

    fun curvatureDrive(power: Double, rotation: Double) {
        mode = Mode.CURVATURE_DRIVE
        setArcadePowers(power, rotation)
    }

    private fun setArcadePowers(power: Double, rotation: Double) {
        rightPower = 0.0
        leftPower = rightPower
        this.power = power
        this.rotation = rotation
    }

    override fun periodic() {
        when (mode) {
            Mode.TANK_DRIVE -> {
                differentialDrive.tankDrive(leftPower, rightPower, false)
            }
            Mode.ARCADE_DRIVE -> {
                differentialDrive.arcadeDrive(power, rotation, false)
            }
            Mode.CURVATURE_DRIVE -> {
                differentialDrive.curvatureDrive(power, rotation, true)
            }
            null -> {}
        }
    }

    enum class Mode {
        TANK_DRIVE,
        CURVATURE_DRIVE,
        ARCADE_DRIVE;
    }
}